src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -325,8 +325,16 @@ public synchronized Object get(long timeout, TimeUnit unit) throws InterruptedEx
 
     private void setOutcome(Outcome o) {
         synchronized (this) {
-            if (bodyToUnexport != null && thread != null) {
-                thread.group.unexport(bodyToUnexport);
+            if (thread != null) {
+                if (bodyToUnexport != null) {
+                    thread.group.unexport(bodyToUnexport);
+                }
+                // onSuccess and onFailure are part of the program and reference the CpsBodyExecution as their outer
+                // class. They are usually unreachable from the program once the body completes, but in some cases they
+                // may still be reachable (e.g. closures that outlive the body). If the CpsBodyExecution is still part
+                // of the program, we need to make sure that we do not reference to a dead CpsThread because that may
+                // cause resumption issues (JENKINS-63164).
+                thread = null;
```

Comment:
I'm not sure about this. Some other ideas I had:

* Update `CpsThreadGroup.run` so that when it removes a thread from `CpsThreadGroup.threads` it calls some new method `CpsThread.destroy` or similar that nulls out fields (mainly `contextVariables`) in `CpsThread` in case something is still referencing it. Might fix other issues as well that we aren't aware of.
* Modify `groovy-cps` so that the `Env` captured by `CpsClosure` does not reference things like `CallEnv.returnAddress` from nested environments. Not sure about correctness/feasibility of this, seems complicated. At least in the simple case here we only need the locals from the nested environments and if we only stored those variables there would be no issues, but perhaps there could be a references from the locals to some of the fields in `CallEnv`, and maybe we need to maintain the actual `Env` references as-is for semantics.

Reply:
The first option sounds better to me.  Is there a downside to that?

Reply:
> The first option sounds better to me. Is there a downside to that?

I'm not really sure, I can prototype it. If `CpsBodyExecution.thread` is actually usable as a captured variable in the closure, then nulling out all of the fields in `CpsThread` could result in some weird behavior/NPEs, but I suspect that the `CpsThread` is not actually usable.

Reply:
Noting some other ideas that came up in a discussion today in case they are useful to someone in the future:

* Replacing `CpsBodyExecution.thread` with `CpsBodyExecution.threadId` and only holding a transient reference to the `CpsThread` itself, loading the value by looking up the `threadId` in `CpsThreadGroup`. Not sure if this would be possible, and would require some adjustments to maintain serialization compatiblity.
* Hooking into `RiverWriter` or similar code in `workflow-support` and just not serializing `CpsThread`s that are not part of the root `CpsThreadGroup`. Seems complicated, and I'm not sure about compatibility.